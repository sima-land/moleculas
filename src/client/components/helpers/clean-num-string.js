import pipe from 'lodash/fp/pipe';
import toString from 'lodash/toString';

/**
 * Очищает строку от любых символов, кроме цифр и удаляет лишние нули вначале.
 * @param {string} str Строка.
 * @return {string} Возвращает очищенную строку.
 */
export const cleanNumString = pipe(
  str => toString(str) && str.replace(/[^0-9]/g, ''),
  str => str === '' ? '' : Number(str).toString()
);
