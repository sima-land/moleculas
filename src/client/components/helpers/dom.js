import isNumber from 'lodash/isNumber';

/**
 * Обрезает textContent переданного элемента, пока его высота не будет соответствовать заданной.
 * @param {HTMLElement} element Элемент для обрезания textContent.
 * @param {number} height Заданная высота.
 */
export const cutTextContent = (element, height) => {
  if (element instanceof HTMLElement && isNumber(height) && height > 0) {
    const selfDisplay = element.style.display;

    element.style.display = 'block';
    while (element.textContent.length > 3 && element.clientHeight > height) {
      // удаляем 4 элемента и вставляем многоточие
      element.textContent = `${element.textContent.slice(0, -4).trim()}...`;
    }
    element.style.display = selfDisplay;
  }
};
