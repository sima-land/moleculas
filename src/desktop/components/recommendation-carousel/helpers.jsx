import { useEffect, useRef, useState } from 'react';
import on from '@dev-dep/ui-nucleons/helpers/on';

/**
 * Управляет расположением и появлением тултипа.
 * @return {Object} Объект с рефом и функциями для управления тултипом.
 */
export const useControlTooltip = () => {
  const [canShowTooltip, setCanShowTooltip] = useState(true);
  const tooltipRef = useRef();

  /**
   * Скрывает тултип.
   */
  const hideTooltip = () => {
    if (tooltipRef.current) {
      tooltipRef.current.style.opacity = 0;
    }
  };

  useEffect(() => on(document, 'mousedown', () => {
    setCanShowTooltip(false);
    hideTooltip();
  }), []);

  useEffect(() => on(document, 'mouseup', () => setCanShowTooltip(true)), []);

  /**
   * Происходит при наведении на кнопку быстрого просмотра.
   * @param {Event} event Событие при наведении на кнопку быстрого просмотра.
   */
  const onHoverQuickViewButton = event => {
    if (canShowTooltip && tooltipRef.current) {
      const { left, top } = event.currentTarget.getBoundingClientRect();
      tooltipRef.current.style.right = `calc(100% - ${left - 8}px)`; // 8px - отступ от иконки.
      tooltipRef.current.style.top = `${top + window.scrollY + 12}px`; // 12px - половина никонки.
      tooltipRef.current.style.opacity = 1;
    }
  };

  return {
    tooltipRef,
    onHoverQuickViewButton,
    onLeaveQuickViewButton: hideTooltip,
  };
};
