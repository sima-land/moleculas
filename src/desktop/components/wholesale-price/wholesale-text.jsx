import React, { Component, createRef } from 'react';
import isFunction from 'lodash/isFunction';
import withGlobalListeners from '@dev-dep/ui-nucleons/hoc/with-global-listeners';
import PositioningPopup from '@dev-dep/ui-nucleons/popups/positioning-popup';
import classnames from 'classnames/bind';
import style from './wholesale-price.scss';

export const cx = classnames.bind(style);

/**
 * Компонент текста цены крупного опта с описанием.
 * @param {Object} props Свойства компонента.
 * @param {string} [props.description] Описание условий акции крупного опта.
 * @param {string} [props.className] Дополнительный класс компонента.
 * @param {Function} [props.onDetailsClick] Дополнительные действия при клике открытия описания.
 */
export class WholesaleText extends Component {
  state = {
    isPopupOpen: false,
  };
  openerRef = createRef();
  textWrapperRef = createRef();
  removeGlobalListener;

  /**
   * @inheritDoc
   */
  constructor (props) {
    super(props);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
  }

  /**
   * Обработчик клика вне компонента.
   * @param {Event} event Событие клика.
   */
  handleOutsideClick (event) {
    const ref = this.textWrapperRef && this.textWrapperRef.current;
    if (event && event.target && ref && !ref.contains(event.target)) {
      this.setState({ isPopupOpen: false });
    }
  }

  /**
   * @inheritdoc
   */
  componentDidMount () {
    const { addGlobalListener, description } = this.props;
    if (description && isFunction(addGlobalListener)) {
      this.removeGlobalListener = addGlobalListener('click', this.handleOutsideClick);
    }
  }

  /**
   * @inheritdoc
   */
  componentWillUnmount () {
    if (isFunction(this.removeGlobalListener)) {
      this.removeGlobalListener();
    }
  }

  /**
   * @inheritdoc
   */
  render () {
    const {
      description,
      onDetailsClick,
      text,
    } = this.props;
    return (
      <div ref={this.textWrapperRef} className={cx('wholesale-text-wrapper')}>
        <span
          className={cx('wholesale-text')}
          onClick={() => {
            if (!this.state.isPopupOpen) {
              Boolean(description) && isFunction(onDetailsClick) && onDetailsClick();
            }
            this.setState({ isPopupOpen: !this.state.isPopupOpen });
          }}
          ref={this.openerRef}
          role={Boolean(description) && isFunction(onDetailsClick) ? 'button' : 'note'}
          children={text}
        />
        {Boolean(description) && this.state.isPopupOpen && (
          <PositioningPopup
            withArrow
            className={cx('popup')}
            opener={this.openerRef}
            parent={this.textWrapperRef}
            horizontalPosition='center'
            basePopupClass={cx('notice-popup')}
          >
            <div className={cx('popup-body')} dangerouslySetInnerHTML={{ __html: description }} />
          </PositioningPopup>
        )}
      </div>
    );
  }
}

export default withGlobalListeners(WholesaleText);
