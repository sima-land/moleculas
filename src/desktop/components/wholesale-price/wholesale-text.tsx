import React, { Component, createRef } from 'react';
import isFunction from 'lodash/isFunction';
import on from '@dev-dep/ui-nucleons/helpers/on';
import PositioningPopup from '../__temp__/popups/positioning-popup';
import classnames from 'classnames/bind';
import style from './wholesale-price.scss';

export interface WholesaleTextProps {
  description: string
  onDetailsClick?: () => void
  text?: string
}

export const cx = classnames.bind(style);

/**
 * Компонент текста цены крупного опта с описанием.
 */
class WholesaleText extends Component<WholesaleTextProps, any> {
  state = {
    isPopupOpen: false,
  };
  openerRef = createRef<HTMLSpanElement>();
  textWrapperRef = createRef<HTMLDivElement>();
  removeGlobalListener: any;

  /**
   * @inheritDoc
   */
  constructor (props: WholesaleTextProps) {
    super(props);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
  }

  /**
   * Обработчик клика вне компонента.
   * @param {Event} event Событие клика.
   */
  handleOutsideClick (event: any) {
    const ref = this.textWrapperRef && this.textWrapperRef.current;
    if (event && event.target && ref && !(ref as any).contains(event.target)) {
      this.setState({ isPopupOpen: false });
    }
  }

  /**
   * @inheritdoc
   */
  componentDidMount () {
    const { description } = this.props;

    if (description) {
      this.removeGlobalListener = on(window, 'click', this.handleOutsideClick);
    }
  }

  /**
   * @inheritdoc
   */
  componentWillUnmount () {
    if (isFunction(this.removeGlobalListener)) {
      this.removeGlobalListener();
    }
  }

  /**
   * @inheritdoc
   */
  render () {
    const { description, onDetailsClick, text } = this.props;
    return (

      <div ref={this.textWrapperRef} className={cx('wholesale-text-wrapper')}>
        <span className={cx('wholesale-text')}
          onClick={() => {
            if (!this.state.isPopupOpen) {
              Boolean(description) && isFunction(onDetailsClick) && onDetailsClick();
            }
            this.setState({ isPopupOpen: !this.state.isPopupOpen });
          }}
          ref={this.openerRef}
          role={Boolean(description) && isFunction(onDetailsClick) ? 'button' : 'note'}
          children={text}
        />
        {Boolean(description) && this.state.isPopupOpen && (
          <PositioningPopup
            withArrow
            className={cx('popup')}
            opener={this.openerRef}
            parent={this.textWrapperRef}
            horizontalPosition='center'
            basePopupClass={cx('notice-popup')}
          >
            <div className={cx('popup-body')} dangerouslySetInnerHTML={{ __html: description }} />
          </PositioningPopup>
        )}
      </div>
    );
  }
}
export default WholesaleText;
