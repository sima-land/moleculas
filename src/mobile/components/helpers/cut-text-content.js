import isNumber from 'lodash/isNumber';

/**
 * Обрезает textContent переданного элемента, пока его высота не будет соответствовать заданной.
 * @param {HTMLElement} element Элемент для обрезания textContent.
 * @param {number} height Заданная высота.
 */
export const cutTextContent = (element, height) => {
  if (element instanceof HTMLElement && isNumber(height) && height > 0) {
    const selfDisplay = element.style.display;
    const minLength = 15;

    element.style.display = 'block';
    if (element.textContent.length > minLength && element.clientHeight > height) {
      const capacityLength = element.textContent.length / (element.clientHeight / height);

      // удаляем 4 символа и вставляем многоточие
      element.textContent = `${element.textContent.slice(0, capacityLength - 4).trim()}...`;
    }
    element.style.display = selfDisplay;
  }
};
