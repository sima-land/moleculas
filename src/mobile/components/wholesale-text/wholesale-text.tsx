import React from 'react';
import classnames from 'classnames';
import styles from './wholesale-text.module.scss';

/**
 * Компонент текста крупного опта.
 * @param props Свойства компонента.
 * @param props.text Текст крупного опта.
 * @param props.className Дополнительный класс компонента.
 * @return Элемент.
 */
const WholesaleText = ({
  text,
  className,
}: {
  text?: string
  className?: string
}) => (
  <span className={classnames(styles.root, className)}>{text}</span>
);

export default WholesaleText;
