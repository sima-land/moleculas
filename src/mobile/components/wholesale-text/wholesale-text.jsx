import React, { useRef, useEffect } from 'react';
import debounce from 'lodash/debounce';
import withGlobalListeners from '@dev-dep/ui-nucleons/hoc/with-global-listeners';
import { cutTextContent } from '../helpers/cut-text-content';

/**
 * @type {number} Максимальная высота текста крупного опта.
 */
const linkMaxHeight = 38;

/**
 * Обрезает текст до нужной высоты.
 * @param {HTMLElement} element Элемент с текстом.
 */
export const cutItemWholesaleText = element => {
  cutTextContent(element, linkMaxHeight);
};

/**
 * Компонент текста крупного опта.
 * @param {Object} props Свойства компонента.
 * @param {string} props.text Текст крупного опта.
 * @param {string} [props.className] Дополнительный класс компонента.
 * @return {ReactElement} Компонент цены крупного опта с описанием.
 */
export const WholesaleText = ({
  text,
  className,
  addGlobalListener,
}) => {
  const linkRef = useRef(null);

  useEffect(() => {
    const { current: element } = linkRef;
    element && cutItemWholesaleText(element);

    const cutItemText = debounce(() => {
      element.textContent = text;
      cutItemWholesaleText(element);
    }, 500);

    return addGlobalListener('resize', cutItemText);
  }, [text]);

  return (
    <span className={className} ref={linkRef}>{text}</span>
  );
};

export default withGlobalListeners(WholesaleText);
