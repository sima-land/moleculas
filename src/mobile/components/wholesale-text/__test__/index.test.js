import React from 'react';
import { mount } from 'enzyme';
import { WholesaleText } from '../wholesale-text';
import { cutTextContent } from '../../helpers/cut-text-content';
import debounce from 'lodash/debounce';

jest.mock('lodash/debounce', () => jest.fn(fn => fn));

jest.mock('../../helpers/cut-text-content', () => {
  const original = jest.requireActual('../../helpers/cut-text-content');

  return {
    ...original,
    __esModule: true,
    cutTextContent: jest.fn(original.cutTextContent),
  };
});

describe('<WholesaleText />', () => {
  const addGlobalListener = jest.fn();
  beforeEach(() => {
    addGlobalListener.mockClear();
  });

  it('should render without props', () => {
    const wrapper = mount(<WholesaleText addGlobalListener={addGlobalListener} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('should render with props and cutTextContent calls correctly', () => {
    const wrapper = mount(
      <WholesaleText
        text='Крупный опт «Галантерея и швейная галантерея»'
        addGlobalListener={addGlobalListener}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(cutTextContent).toHaveBeenCalled();

    cutTextContent.mockClear();
    expect(cutTextContent).toHaveBeenCalledTimes(0);

    expect(debounce).toHaveBeenCalled();
    expect(addGlobalListener).toHaveBeenCalledTimes(1);
    expect(addGlobalListener.mock.calls[0][0]).toEqual('resize');
    const callback = addGlobalListener.mock.calls[0][1];
    callback();

    expect(cutTextContent).toHaveBeenCalledTimes(1);
  });
});
