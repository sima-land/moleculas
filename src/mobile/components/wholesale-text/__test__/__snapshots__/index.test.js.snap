// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WholesaleText /> should render with props and cutTextContent calls correctly 1`] = `
<WholesaleText
  text="Крупный опт «Галантерея и швейная галантерея»"
>
  <span
    className="root"
  >
    Крупный опт «Галантерея и швейная галантерея»
  </span>
</WholesaleText>
`;
